linking/preload/buffered.go

```diff
@@ -0,0 +1,210 @@
+package preload
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/linking"
+)
+
+type preloadingLink struct {
+	refCnt     uint64
+	loadSyncer sync.Once
+	loaded     chan struct{}
+	data       []byte
+	err        error
+}
+
+type request struct {
+	lnkCtx linking.LinkContext
+	lnk    datamodel.Link
+}
+
+type BufferedLoader struct {
```

Comment:
doc comment for a buffered loader would be useful. - in particular i  initially just understood it as pre-loading links in the traversal, but the set of optimizations around the byteReader caught me off guard, as that seems like a potentially different concern.

Reply:
Currently looking at either ripping this out of this PR or implementing it in more of a shell such that the workers and data stores/sources have to all be plugged in and all you get is the core queueing and shunting logic.

It certainly is quite separate to the PR itself, which is pretty good as it is and doesn't need this, it's just going to be a common pattern. I think the best approach might be to implement it elsewhere and then pull back the core logic in when it's been proven/discovered.